Introduce `monotonicity` parameter to the modules

ybubnov opened the pull request:
This patch does introduces monotonicity parameter (non-decreasing or non-increasing), which enforces the monotoicity of the control points in the grids. This fine control allows to yield non-linear monotonic dependencies. By default behaviour is preserved and monotonicity is not enforced.

Also this patch updates formatting/linting: ruff version is updated, black is dropped (they are doing the same and override output of each other, which is very annoying). The preference is given to the ruff since it was the first in the `.pre-commit-config.yaml`.

ybubnov committed:
Introduce `monotonicity` parameter to the modules

This patch does introduces monotonicity parameter (non-decreasing or
non-increasing), which enforces the monotoicity of the control points
in the grids. This fine control allows to yield non-linear monotonic
dependencies. By default behaviour is preserved and monotonicity is
not enforced.

Also this patch updates formatting/linting: ruff version is updated,
black is dropped (they are doing the same and override output of each
other, which is very annoying). The preference is given to the ruff
since it was the first in the `.pre-commit-config.yaml`.

ybubnov commented:
@alisterburt unfortunately, I cannot assign reviewers, so referring you personally here. I would appreciate getting a review on the proposed changes. Thanks!

alisterburt commented on src/torch_cubic_spline_grids/_base_cubic_grid.py:
```diff
@@ -21,33 +23,40 @@ def __init__(
-            self._interpolation_function(
-                self._data, minibatch_u, matrix=self.interpolation_matrix
-            )
+            interpolation_function(minibatch_u)
```
nice, clean!

alisterburt reviewed (commented):
Hi @ybubnov - thanks for taking the time to open a PR, glad the package was useful for you!

Overall this looks good and I'm happy to merge, a few small questions:
- is it standard to describe monotonicity as 'nonincreasing' and 'nondecreasing'?
- would it be better to type the parameter as `monotonicity: Literal['nonincreasing', 'nondecreasing']`

Describing monotonicity as nonincreasing or nondecreasing feels like a double negative to me and maybe 'increasing' and 'decreasing' would be simpler? I'm not familiar with why this feature would be desired though so please do tell me if I'm off base

alisterburt commented:
also thanks for the general code quality improvements, much appreciated!

alisterburt commented:
there's also a weird test failure only on py38 https://github.com/teamtomo/torch-cubic-spline-grids/actions/runs/14449832407/job/40543142291?pr=29

ybubnov committed:
Address comments: use literal to specify monotonicity

This patch also enforces monotonicity of the grid itself, instead of
control points.

ybubnov commented:
Thank you for the quick turn around and review comments, appreciate that.

> is it standard to describe monotonicity as 'nonincreasing' and 'nondecreasing'?

I'm trying to use cubic splines to fit Generalized Additive Model. Such types of models are quite good at modelling the non-linear dependencies between predictors. In python world, there are very few packages that are doing that, but it's quite popular in R. In R the monotonicity is rather specified through "constraint" on a shape of the spline. For example in a very popular `mgcv`, [the monotonicity constraint](https://stat.ethz.ch/R-manual/R-devel/library/mgcv/html/mono.con.html) is specified as `mono.con(up = TRUE)` for increasing monotonicity and `mono.con(up = FALSE)` for decreasing monotonicity. But on my personal taste, I find `monotonicity="increasing"` as a more verbose and self-documenting approach. In general, I'm open to the changes.

> would it be better to type the parameter as monotonicity: Literal['nonincreasing', 'nondecreasing']

Agree, replaced with `MonotonicityType`, which is a literal as `Literal['increasing', 'decreasing']`.

> Describing monotonicity as nonincreasing or nondecreasing feels like a double negative to me and maybe 'increasing' and 'decreasing' would be simpler?

Agree, in Wikipedia both terminologies are used, so 'increasing' and 'decreasing' indeed are bringing less mental load.

> I'm not familiar with why this feature would be desired though so please do tell me if I'm off base.

In general, I need to build a regression model, but some dependencies between target variable and predictors are non-linear. In fact, some of those non-linear dependencies cannot be decreasing. Let's say, as examples: the dependency between gasoline spend and speed (speed cannot decrease, when more gasoline is added, it could stay at max, but never decrease), or a volume of purchases, once the delivery distance from a venue is increased (once you increase delivery radius, the volume of purchases never decreases).

alisterburt reviewed (approved):
@ybubnov thank you for the work you've put into this! Looks good to me, I'll merge and cut a release now

alisterburt commented:
deploy workflow here: https://github.com/teamtomo/torch-cubic-spline-grids/actions/runs/14503838613

I've added you as a maintainer here in case you want to help out with maintenance here, it's been fun working with you 🙂 if you accept let's please still discuss larger things over issues/PRs
